### ISSUE TITLE
Error CS0246 instead of CS1069 for forwarded type in unreferenced assembly

### ISSUE THREAD

**Issue by joncham** [dotnet/roslyn#29747]
**Version Used**: 
```
Microsoft (R) Visual C# Compiler version 2.7.0.65535 (6e7d8658)
Copyright (C) Microsoft Corporation. All rights reserved.
```

**Steps to Reproduce**:

1. Compile source code which references an assembly with a forwarded type, but not referencing the assembly to which the type is forwarded. Filing this after talking to @agocke . There was apparently an old bug on this in a different tracking system. He recommended I paste that old text here:

```
Suppose a.dll forwards "Namespace.Forwarded" to b.dll and the following source is compiled with a reference to a.dll, but no reference to b.dll.

 

using Namespace;

class Test

{

static void Main()

    {

        var f = new Forwarded();

    }

}

 

We would like to report CS1069 - type forwarded to b.dll, but we actually report CS0246 - type not found.  This happens because we assume Forwarded is the full metadata name of the type, but it's actually Namespace.Forwarded.  (If it was qualified as Namespace.Forwarded in source, or if it was in the global namespace, we wouldn't have this problem.)

 

Option 1: Include type forwarders in lookup results.

Option 2: Try all possible usings and containing namespaces.

Option 3: Look for a metadata name with that name as a suffix.  To be totally correct, we'd have to confirm that we could look up the part that doesn't appear in source.

Option 4: Do nothing - suboptimal diagnostics are not an urgent problem.
``` 

**Expected Behavior**:
CS1069 

**Actual Behavior**:
CS0246 - @agocke indicated this is because option 4 above was chosen above.

**Comment by jechter** [comment]
Some context: In Unity, we have switched our public APIs to a modular setup with different modules in different dlls - previously it was all in one UnityEngine.dll. Now, we still generate UnityEngine.dll with type forwarders to the new API locations, so old assemblies referencing UnityEngine.dll continue to work. We would also pass this forwarder UnityEngine.dll to the compiler for user code, so that if some modules are not used, we can get meaningful compiler errors (CS1069), telling the user which module they need to include in the project. This works fine in mcs. But in csc, all we get is CS0246 - type not found, which is not as helpful.

**Comment by agocke** [comment]
One more thing that I got from the discussion on the internal bug: (4) was chosen at the time because our telemetry said that this error had *never been reported in a production release of Visual Studio*. This calculus is obviously very different now, so we should reconsider the options (which I think are the same).

**Comment by jechter** [comment]
So, because I get asked about this issue multiple times a week at the moment, I decided to take a look myself.

I looked at the Roslyn code, and came up with the patch below to make it return the correct error code. Now, obviously, I don't have a clue about the Roslyn code base, and this is probably a bad, hackish way of fixing it. Basically, what it does is to make `Imports.LookupSymbolInUsings` also check any referenced assemblies for TypeForwarders of that name, and merge them to `results` with an  `ERR_DottedTypeNameNotFoundInNSFwd` error. Then, in `Binder.ResultSymbol`, it will skip the call to `ReportUseSiteDiagnostics` if there is an existing error attached to the symbol, so we don't get duplicate errors.

Does this make any sense at all? Any suggestions for improving my fix?

```
diff --git a/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs b/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
index da9092875f..185c889787 100644
--- a/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
@@ -1642,7 +1642,10 @@ public int Compare(Symbol fst, Symbol snd)
 
                         if (!suppressUseSiteDiagnostics)
                         {
-                            wasError = ReportUseSiteDiagnostics(singleResult, diagnostics, where);
+                            if (result.Error != null)
+                                wasError = true;
+                            else
+                                wasError = ReportUseSiteDiagnostics(singleResult, diagnostics, where);
                         }
                         else if (singleResult.Kind == SymbolKind.ErrorType)
                         {
@@ -1692,7 +1695,7 @@ public int Compare(Symbol fst, Symbol snd)
             Debug.Assert(symbols.Count > 0);
 
             // Report any errors we encountered with the symbol we looked up.
-            if (!suppressUseSiteDiagnostics)
+            if (!suppressUseSiteDiagnostics && result.Error == null)
             {
                 for (int i = 0; i < symbols.Count; i++)
                 {
diff --git a/src/Compilers/CSharp/Portable/Binder/Imports.cs b/src/Compilers/CSharp/Portable/Binder/Imports.cs
index c28c1b2626..c357c92055 100644
--- a/src/Compilers/CSharp/Portable/Binder/Imports.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Imports.cs
@@ -706,6 +706,22 @@ internal bool IsUsingAlias(string name, bool callerIsSemanticModel)
 
                     result.MergeEqual(res);
                 }
+
+                var fullName = typeOrNamespace.NamespaceOrType + "." + name;
+                var metadataName = MetadataTypeName.FromFullName(fullName);
+                var containingAssembly = originalBinder.Compilation.Assembly;
+
+                NamedTypeSymbol forwardedType = null;
+                foreach (var referencedAssembly in containingAssembly.Modules[0].GetReferencedAssemblySymbols())
+                {
+                    forwardedType = referencedAssembly.TryLookupForwardedMetadataType(ref metadataName);
+                    if ((object)forwardedType != null)
+                    {
+                        var diagInfo = new CSDiagnosticInfo(ErrorCode.ERR_DottedTypeNameNotFoundInNSFwd, name, typeOrNamespace.NamespaceOrType, forwardedType.ContainingAssembly);
+                        result.MergeEqual(LookupResult.NotReferencable(forwardedType, diagInfo));
+                    }
+                }
+
             }
         }
```

**Comment by agocke** [comment]
@jechter I would start by creating a PR. It will be easier to leave comments and it'll also see if your change passes tests.

**Comment by jcouv** [comment]
@agocke The linked PR (#30244) was merged. Can this issue be closed? If not, please push to Compiler.Next

**Comment by agocke** [comment]
Looks like this was fixed in #30244
